atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java

```diff
@@ -362,31 +371,103 @@ public void putWithTimestamps(TableReference tableRef, Multimap<Cell, Value> val
+            Key key = getKey(table, e.getKey(), timestamp);
+            byte[] oldContents = putIfAbsent(table, key, contents);
             if (oldContents != null && (doNotOverwriteWithSameValue || !Arrays.equals(oldContents, contents))) {
                 throw new KeyAlreadyExistsException("We already have a value for this timestamp");
             }
         }
     }
 
+    // Returns the existing contents, if any, and null otherwise
+    private byte[] putIfAbsent(Table table, Key key, final byte[] contents) {
+        return table.entries.putIfAbsent(key, copyOf(contents));
+    }
+
     @Override
-    public void putUnlessExists(TableReference tableRef, Map<Cell, byte[]> values)
-            throws KeyAlreadyExistsException {
-        putInternal(tableRef, KeyValueServices.toConstantTimestampValues(values.entrySet(), 0), true);
+    public void checkAndSet(CheckAndSetRequest request) throws CheckAndSetException {
+        Table table = getTableMap(request.table());
+        Cell cell = request.cell();
+        Optional<byte[]> oldValue = request.oldValue();
+        byte[] contents = request.newValue();
+
+        Key key = getKey(table, cell, AtlasDbConstants.TRANSACTION_TS);
+        if (oldValue.isPresent()) {
+            byte[] storedValue = table.entries.get(key);
+            boolean succeeded = Arrays.equals(storedValue, oldValue.get())
+                    && table.entries.replace(key, storedValue, copyOf(contents));
+            if (!succeeded) {
+                byte[] actual = table.entries.get(key); // Refetch, something may have happened between get and replace
+                throwCheckAndSetException(cell, oldValue.get(), actual);
+            }
+        } else {
+            byte[] oldContents = putIfAbsent(table, key, contents);
+            if (oldContents != null) {
+                String actual = new String(oldContents, StandardCharsets.UTF_8);
+                String msg = String.format("Unexpected value for this key. Wanted (empty), got %s", actual);
+                throw new CheckAndSetException(msg, cell, null, ImmutableList.of(oldContents));
+            }
+        }
+    }
+
+    private void throwCheckAndSetException(Cell cell, byte[] expected, byte[] actual) {
+        String expectedStr = new String(expected, StandardCharsets.UTF_8);
+        if (actual != null) {
+            String actualStr = new String(actual, StandardCharsets.UTF_8);
+            String msg = String.format("Unexpected value for this key. Wanted %s, got %s", expectedStr, actualStr);
+            throw new CheckAndSetException(msg, cell, expected, ImmutableList.of(actual));
+        } else {
+            String msg = String.format("Unexpected value for this key. Wanted %s, got (empty)", expectedStr);
+            throw new CheckAndSetException(msg, cell, expected, ImmutableList.of());
+        }
+    }
+
+    private List<byte[]> getStoredValues(Cell cell, TableReference tableRef) {
```

Comment:
Yep - they're a hangover from an old version of the impl, which I forgot to remove once I'd made the tests pass without them. (Ditto below)